From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DoctaEnkoda <bierquejason@gmail.com>
Date: Tue, 3 Jan 2023 03:35:42 +0100
Subject: [PATCH] Event When Player Join Leave

Add PlayerLeaveExternalServerEvent when Player Leave an External Server
Add PlayerJoinExternalServerEvent when Player Join an External Server

Beware, this does not replace PlayerJoinEvent and PlayerQuitEvent but still allows to know when a player leaves a server on another instance.

diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/PlayerCreatePacket.java b/src/main/java/puregero/multipaper/externalserverprotocol/PlayerCreatePacket.java
index 7fc4a78b3b6138c662abc12b8ce825f09a9fd26f..ae108bf320a624b3922bafb386857a83937994e3 100644
--- a/src/main/java/puregero/multipaper/externalserverprotocol/PlayerCreatePacket.java
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/PlayerCreatePacket.java
@@ -12,10 +12,12 @@ import net.minecraft.world.entity.HumanoidArm;
 import net.minecraft.world.level.GameType;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.bukkit.Bukkit;
 import org.bukkit.event.player.PlayerKickEvent;
 
 import puregero.multipaper.*;
 import puregero.multipaper.config.MultiPaperConfiguration;
+import puregero.multipaper.event.player.PlayerJoinExternalServerEvent;
 
 import java.net.InetAddress;
 import java.net.InetSocketAddress;
@@ -147,6 +149,8 @@ public class PlayerCreatePacket extends ExternalServerPacket {
             ExternalPlayer player = ExternalPlayer.create(connection, gameProfile, world, x, y, z, yaw, pitch, gamemode, address, advancements, stats, entityId);
             player.getBukkitEntity().data = data;
             player.getBukkitEntity().persistentData = persistentData;
+            PlayerJoinExternalServerEvent playerJoinExternalServerEvent = new PlayerJoinExternalServerEvent(gameProfile.getId(), gameProfile.getName(), MultiPaperConfiguration.get().masterConnection.myName);
+            Bukkit.getPluginManager().callEvent(playerJoinExternalServerEvent);
         });
     }
 
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/PlayerRemovePacket.java b/src/main/java/puregero/multipaper/externalserverprotocol/PlayerRemovePacket.java
index f37bdc4eb9cd13732ce2dc2848401b80fa48d88f..a14f2afe5dff010069ea7569aedb18ada347a90e 100644
--- a/src/main/java/puregero/multipaper/externalserverprotocol/PlayerRemovePacket.java
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/PlayerRemovePacket.java
@@ -6,10 +6,12 @@ import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.level.ServerPlayer;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.bukkit.Bukkit;
 import org.bukkit.event.player.PlayerKickEvent;
 import puregero.multipaper.ExternalServerConnection;
 import puregero.multipaper.MultiPaper;
 import puregero.multipaper.config.MultiPaperConfiguration;
+import puregero.multipaper.event.player.PlayerLeaveExternalServerEvent;
 
 import java.util.UUID;
 
@@ -45,6 +47,8 @@ public class PlayerRemovePacket extends ExternalServerPacket {
             }
 
             player.connection.disconnect(EXTERNAL_DISCONNECT_COMPONENT, PlayerKickEvent.Cause.TIMEOUT);
+            PlayerLeaveExternalServerEvent playerLeaveExternalServerEvent = new PlayerLeaveExternalServerEvent(player.getGameProfile().getId(), player.getGameProfile().getName(), MultiPaperConfiguration.get().masterConnection.myName);
+            Bukkit.getPluginManager().callEvent(playerLeaveExternalServerEvent);
         });
     }
 }
